### DISCUSSION TITLE
useService performance recommendations

### DISCUSSION THREAD

**Discussion by alexreardon** [statelyai/xstate#1209] [General]
I'm trying to find a good way to use xstate at scale with good performance.

I create a machine fairly high in my react tree

```js
const [state, send, service] = useMachine(machine);
```

I put the `service` on the react `<ServiceContext.Provider value={service}>` so that other components can also talk to the service

```js
const [state, send, service] = useService(service);
```

My trouble is that in order for `state` to be updated in everything with `useService`, they are re-rendered.

```js
const [state, send, service] = useService(service);
//     ^^^^^
```

I often have components that care about some state transitions, but not others. But right now each one has to render every time.

I'm looking for a good way to solve this.

### Some ideas:

1. Use more nested machines (in user land)
2. Create some sort of selector mechanism so I only re-render components that need to (in user land). This would involve creating a wrapping component for each useService which does some memoization
3. Something else?

**Comment by davidkpiano** [comment]
I'm thinking something like this can help for now:

```jsx
function usePartial(service, selector) {
  const [partialContext, setPartialContext] = useState(selector(service.state.context));

  useEffect(() => {
    const { unsubscribe } = service.subscribe(state => {
      if (!state.changed) return;

      setPartialContext(selector(state.context));
    });

    return () => unsubscribe();
  }, []);

  return partialContext;
}
```

**Comment by alexreardon** [comment]
So:

> 2. Create some sort of selector mechanism

👍

**Comment by JackCA** [comment]
I've written a slightly modified typescript version that uses `state` instead of `context` as the argument:

``` typescript
export function usePartial<
  TContext,
  TSchema,
  TEvent extends EventObject,
  TSelectorValue
>(
  service: Interpreter<TContext, TSchema, TEvent>,
  selector: (state: State<TContext, TEvent, TSchema>) => TSelectorValue
): TSelectorValue {
  const [partialState, setPartialState] = useState(selector(service.state))

  useEffect(() => {
    const { unsubscribe } = service.subscribe(state => {
      if (state.changed !== true) return

      setPartialState(selector(state))
    })

    return (): void => unsubscribe()
  }, [selector, service])

  return partialState
}
```

**Comment by nazmifeeroz** [comment]
this code is interesting and looks legit.
I tried it out, but i still getting re-renders from other child components not related to the context changed.

it seems that anything under the same context provider still gets re-rendered.

maybe i missed something out?
do you mind creating a sandbox i could refer to?

**Comment by davidkpiano** [comment]
@nazmifeeroz Chances are the rerenders are _not_ due to XState, but other things in React. Try using `React.memo(...)`.

**Reply by nazmifeeroz** [reply]
oh its definitely not XState. its just one of those React nuances to figure out to make our React app performant.. using `memo` might be the way to go. Thanks @davidkpiano !

**Comment by JackCA** [comment]
is the optional [`getSnapshot` arg for `useActor` ](https://github.com/davidkpiano/xstate/tree/master/packages/xstate-react#useactoractor-getsnapshot)doing this now?

**Reply by davidkpiano** [reply]
~It actually can be used in that way, yes!~ Nevermind, spoke too soon, it can't.

**Reply by Andarist** [reply]
I don't think that's actually true. `getSnapshot` is only used to read the "initial state" but later on we pump whatever has been given to us to `setState`:
https://github.com/davidkpiano/xstate/blob/cfbe5b273aaee047fb85646e5d1830e4778281bb/packages/xstate-react/src/useActor.ts#L63

There is currently no way to filter those emits.

**Reply by davidkpiano** [reply]
You technically can if it's the exact same value; IIRC a rerender won't happen. 

So it's "fake" filtering, but still effectively the same, depending on goals/use-case.

**Reply by Andarist** [reply]
> You technically can if it's the exact same value; IIRC a rerender won't happen.

Yes, if the value is the same (`===`) the rerender won't happen.

> So it's "fake" filtering, but still effectively the same, depending on goals/use-case.

From what I understand the issue here and the question asked by @JackCA is about a selector mechanism for `useService` and to that I'm referring - `getSnapshot` doesn't try to solve this problem.

**Reply by JackCA** [reply]
My goal is to prevent re-renders unless a specific state-based selector value has changed. In my case, `===` would work (I don't need to do a deep equality check).

@Andarist, I'm a bit confused about:
> getSnapshot is only used to read the "initial state" but later on we pump whatever has been given to us to setState

**Reply by Andarist** [reply]
There is no actual relation between this:
https://github.com/davidkpiano/xstate/blob/cfbe5b273aaee047fb85646e5d1830e4778281bb/packages/xstate-react/src/useActor.ts#L45
and this:
https://github.com/davidkpiano/xstate/blob/cfbe5b273aaee047fb85646e5d1830e4778281bb/packages/xstate-react/src/useActor.ts#L63

Consider `actorRef` being an XState service (created by the `interpret`) and this:
```js
const [state, send] = useActor(service, actor => actor.state.context.foo)
```

This is not quite allowed at the type-level because the return value of `getSnapshot` has to match the type of the emitted value and thus what is type-safe is this:
```js
const [state, send] = useActor(service, state => actor.state)
```
Not that this `getSnapshot` is the default provided by XState for services.

So if you consider types then you can't really use a selector here to select a part of the state reliably and if you only consider runtime then the "selector" (`getSnapshot`) won't actually be used on the emitted values.

**Reply by JackCA** [reply]
@Andarist great explanation, thank you. 

In this case I'm going to continue to use this [implementation](https://github.com/davidkpiano/xstate/discussions/1209#discussioncomment-23747)

**Comment by RareSecond** [comment]
Is the suggested answer now outdated since we have `useSelector`?
